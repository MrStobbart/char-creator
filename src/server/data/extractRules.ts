import { ruleLocations } from './ruleLocations';
import { config } from '../helpers/config';
import logger from '../helpers/winston';
import axios from 'axios';
import mdtable2json from 'mdtable2json';

interface GitHubResponse {
  type: string;
  encoding: string;
  size: number;
  name: string;
  path: string;
  content: string;
  sha: string;
  url: string;
  git_url: string;
  html_url: string;
  download_url: string;
  _links: {
    git: string;
    self: string;
    html: string;
  };
}

interface GitHubErrorResponse {
  message: string;
}

const getRulesFromGithub = async (fileUrl: string) => {
  try {
    const response = await axios.get<GitHubResponse & GitHubErrorResponse>(fileUrl, {
      headers: {
        Authorization: `token ${config.githubToken}`,
      },
    });

    return Buffer.from(response.data.content, 'base64').toString();
  } catch (error) {
    logger.error(`Could not fetch and parse rule file from GitHub: ${error.message}`);
    throw error;
  }
};

export const getTableEndIndex = (ruleFile: string, startIndex: number): number => {
  const nextHashIndex = ruleFile.substring(startIndex).indexOf('#', 15);

  const fixedNextHashIndex = nextHashIndex === -1 ? ruleFile.length : nextHashIndex;

  if (ruleFile.charAt(fixedNextHashIndex - 1) === '(') {
    return getTableEndIndex(ruleFile, fixedNextHashIndex + 15);
  }

  return fixedNextHashIndex + startIndex;
};

export const parseTableForGivenHeader = (ruleFile: string, tableHeader: string) => {
  const tableHeaderIndex = ruleFile.indexOf(tableHeader);

  if (tableHeaderIndex === -1) throw new Error(`Table header '${tableHeader}' does not exist in rule file.`);

  const tableEndIndex = getTableEndIndex(ruleFile, tableHeaderIndex);

  const tableExcerpt = ruleFile.substring(tableHeaderIndex, tableEndIndex);
  const fixedTableExcerpt = tableExcerpt.replace(/[*]/g, '');

  return mdtable2json.getTables(fixedTableExcerpt);
};

export const parseMarkdownTable = (tableExcerpt: string) => {};

const getRules = async (ruleSetName: string) => {
  const ruleLocation = ruleLocations[ruleSetName];
  if (!ruleLocation) throw new Error(`No rule set with the name ${ruleSetName} found`);

  const rules = await getRulesFromGithub(ruleLocation.fileUrl);

  const tableExtract = parseTableForGivenHeader(rules, '#### Hintergrundtalente');
};

getRules('savageRun');

/**
 * 1. Start string until next #
 * 2. Only if #does not have a ( in front of it
 */
