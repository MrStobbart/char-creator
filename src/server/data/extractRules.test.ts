import { getTableEndIndex, parseTableForGivenHeader } from './extractRules';
const exampleRulesFile = `
This is some other Stuff

#### Soziale Talente

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | --------------------------- | ------------------------------------------------------------ | -------------------------- |
| *Demütigung* | Anfänger, Einschüchtern W8 | Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen. | Provozieren wiederholbar |

#### Magische Talente

AH (Magier) impliziert auch Schamanen
`;

const exampleTableExcerpt = `
#### Soziale Talente

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | --------------------------- | ------------------------------------------------------------ | -------------------------- |
| *Demütigung* | Anfänger, Einschüchtern W8 | Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen. | Provozieren wiederholbar |

`;

const exampleTableExcerpt2 = `
#### Fertigkeitsstufen

| Fertigkeitsstufe | Würfel | Bedeutung |
|--------------------|-------------|--------------|
| Nicht vorhanden | W4-1 | Ahnungslos |
| 1 | W4 | Anfänger |
| 2 | W6 | Ausgebildet |
| 3 | W8 | Kompetent |
| 4 | W10 | Herausragend |
| 5 | W12 | Profi |
| 6 | WildDie W8 | Legendär |
| 7 | WildDie W10 | |
| 8 | WildDie W12 | |`;

describe('getTableEndIndex', () => {
  it('returns the correct endIndex of a table', () => {
    const lastTableIndex = getTableEndIndex(exampleRulesFile, 27);
    expect(lastTableIndex).toBe(506);
  });
  it.todo('can handle the special case with the (');
});

describe('parseTableForGivenHeader', () => {
  it('returns the object with the data from the table', () => {
    const parsedTable = parseTableForGivenHeader(exampleRulesFile, '#### Soziale Talente');
    expect(parsedTable).toMatchSnapshot();

    const parsedTable2 = parseTableForGivenHeader(exampleTableExcerpt2, '#### Fertigkeitsstufen');
    expect(parsedTable2).toMatchSnapshot();
  });
});
