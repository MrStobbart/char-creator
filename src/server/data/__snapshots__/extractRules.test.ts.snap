// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTableForGivenHeader returns the object with the data from the table 1`] = `
Array [
  Object {
    "headers": Array [
      "Talent",
      "Anforderungen",
      "Beschreibung",
      "Wirkung",
    ],
    "json": Array [
      Object {
        "Anforderungen": "Anfänger, Einschüchtern W8",
        "Beschreibung": "Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen.",
        "Talent": "Demütigung",
        "Wirkung": "Provozieren wiederholbar",
      },
    ],
  },
]
`;

exports[`parseTableForGivenHeader returns the object with the data from the table 2`] = `
Array [
  Object {
    "headers": Array [
      "Fertigkeitsstufe",
      "Würfel",
      "Bedeutung",
    ],
    "json": Array [
      Object {
        "Bedeutung": "Ahnungslos",
        "Fertigkeitsstufe": "Nicht vorhanden",
        "Würfel": "W4-1",
      },
      Object {
        "Bedeutung": "Anfänger",
        "Fertigkeitsstufe": "1",
        "Würfel": "W4",
      },
      Object {
        "Bedeutung": "Ausgebildet",
        "Fertigkeitsstufe": "2",
        "Würfel": "W6",
      },
      Object {
        "Bedeutung": "Kompetent",
        "Fertigkeitsstufe": "3",
        "Würfel": "W8",
      },
      Object {
        "Bedeutung": "Herausragend",
        "Fertigkeitsstufe": "4",
        "Würfel": "W10",
      },
      Object {
        "Bedeutung": "Profi",
        "Fertigkeitsstufe": "5",
        "Würfel": "W12",
      },
      Object {
        "Bedeutung": "Legendär",
        "Fertigkeitsstufe": "6",
        "Würfel": "WildDie W8",
      },
      Object {
        "Bedeutung": "",
        "Fertigkeitsstufe": "7",
        "Würfel": "WildDie W10",
      },
      Object {
        "Bedeutung": "",
        "Fertigkeitsstufe": "8",
        "Würfel": "WildDie W12",
      },
    ],
  },
]
`;
